pcdsdevices/lodcm.py

```diff
@@ -60,6 +71,683 @@ def __init__(self, prefix, *args, **kwargs):
+
+        Parameters
+        ----------
+        material : str, optional
+            Chemical formula. Defaults to `C`.
+        reflection : tuple, optional
+            Reflection of material. E.g.: `(1, 1, 1)`
+
+        Returns
+        -------
+        energy : number
+            Photon energy in keV.
+        """
+        reflection = reflection or self.get_reflection(
+            as_tuple=True, check=True)
+        length = 2 * np.sin(np.deg2rad(self.th1_c.wm())) * diffraction.d_space(
+            material, reflection)
+        if length == 0:
+            return 0
+        # TODO: do i need to devide here at all?
+        return common.wavelength_to_energy(length) / 1000
+
+    def calc_energy(self, energy, material='C', reflection=None):
+        """Calculate the lom geometry."""
+
+        reflection = reflection or self.get_reflection(
+            as_tuple=True, check=True)
+        th, z = diffraction.get_lom_geometry(energy, material, reflection)
+        return (th, z)
+
+    @pseudo_position_argument
+    def forward(self, pseudo_pos):
+        """
+        Calculate a RealPosition from a given PseudoPosition.
+
+        If the pseudo positioner is here at `pseudo_pos`, then this is where
+        my real motor should be.
+
+        Parameters
+        ----------
+        pseudo_pos : PseudoPosition
+            The pseudo position input, a namedtuple.
+
+        Returns
+        -------
+        real_pos : RealPosition
+            The real position output, a namedtuple.
+        """
+        pseudo_pos = self.PseudoPosition(*pseudo_pos)
+        th, z = self.calc_energy(energy=pseudo_pos.energy)
+        return self.RealPosition(th1_c=th,
+                                 th2_c=th,
+                                 z1_c=-z,
+                                 z2_c=z,
+                                 dr=2*th)
+
+    @real_position_argument
+    def inverse(self, real_pos):
+        """
+        Calculate a PseudoPosition from a given RealPosition.
+
+        If the real motor is at this `real_pos`, then this is where my pseudo
+        position should be.
+
+        Parameters
+        ----------
+        real_pos : RealPosition
+            The real position input.
+
+        Returns
+        -------
+        pseudo_pos : PseudoPosition
+            The pseudo position output.
+        """
+        reflection = self.get_reflection(
+            as_tuple=True, check=True)
+        real_pos = self.RealPosition(*real_pos)
+        length = (2 * np.sin(np.deg2rad(real_pos.th1_c))
+                  * diffraction.d_space('C', reflection))
+        if length == 0:
+            # don't bother transforming this
+            # TODO maybe catch error in common.wave.. when send 0
+            return 0
+        # TODO i might not need to devide here by 1000
+        energy = common.wavelength_to_energy(length) / 1000
+        return self.PseudoPosition(energy=energy)
+
+
+class LODCMEnergySi(PseudoPositioner):
```

Comment:
I split the Energy motors into `Si` and `C` motors - because the `energy` motor is dealing with 8 motors in total, and only 4 at the time depending on the material. So when placing these 8 motors into one PseudoPositioner and using only 4 out of them - it will complain that I am not using the other ones.

I think eventually what users would want to use is only one `energy` motor (as of now i have `energy_c` and `energy_si`)- so I could maybe reassign the `energy` motor at `LODCM __init__` time somehow - based on a function that will let me know if it is `C` or `Si` ....
```
if material == 'C':
    self.energy = self.energy_c
elsif material == 'Si':
    self.energy = self.energy_si

```
or maybe I could go back before I split this motor into two, and in there possibly redefine the `_real` motors to only have 4 at the time.... 

Any suggestions here?

Reply:
Before suggesting anything, I'd like to note a few issues:
- There is some prefix mishap in these classes: the `dr` is supposed to be `XPP:MON:MMS:19` and the `th1_c` is supposed to be `XPP:LODCM:TH1:OFF_C`
- The `OffsetIMS` here are not actually motors, they are just pointing to PVs that store a value:
```
$ caget XPP:LODCM:TH1:OFF_C.RTYP
XPP:LODCM:TH1:OFF_C.RTYP       ao
```
That would say something like `ims` or `motor` if it was a motor:
```
$ caget XPP:MON:MMS:18.RTYP
XPP:MON:MMS:18.RTYP            ims
```

Given this info, I think these two classes can be combined into one that checks whether we're at Si or C and then does the calculation appropriately, if I'm understanding correctly.

Reply:
> Before suggesting anything, I'd like to note a few issues:
> 

Thanks for pointing this out - that's correct, I need to change that.
> * There is some prefix mishap in these classes: the `dr` is supposed to be `XPP:MON:MMS:19` and the `th1_c` is supposed to be `XPP:LODCM:TH1:OFF_C`

Ok thank you, I need to revisit the `MotorOffsetBase` class to see what I have done there...
> * The `OffsetIMS` here are not actually motors, they are just pointing to PVs that store a value:
> 
> ```
> $ caget XPP:LODCM:TH1:OFF_C.RTYP
> XPP:LODCM:TH1:OFF_C.RTYP       ao
> ```
> 
> That would say something like `ims` or `motor` if it was a motor:
> 
> ```
> $ caget XPP:MON:MMS:18.RTYP
> XPP:MON:MMS:18.RTYP            ims
> ```
> 
So here I would have to check if is `Si` or `C` at `__init__` time - which does not sound very good, because I don't know in advance what material we're dealing with... or even if we know, we might want to change it after initialization?
The reason for that, is because the `forward` method will complain for not using all the `_real` motors.

Ex:
if `C`:
     use only 4 motors (`C` motors)
if `Si`:
    use the other 4 motors (`Si` motors)

Are we able to re-define what the `_real` motors are in our class after we initialize the class?
> Given this info, I think these two classes can be combined into one that checks whether we're at Si or C and then does the calculation appropriately, if I'm understanding correctly.

Reply:
You use the same 4 motors in both cases. The `OFF_C` constructs are not motors, they just store values.

Reply:
Ohh ok 🤦 , I'd just use the `th1`, `th2` `z1` and `z2` (maybe `dr` as well) these are the motors - I will come back for more questions then - thanks @ZLLentz

Reply:
Ok back to this guy, can't wait to be done with it. 😓 

Now, if I understand  correctly (probably not yet) our `OffsetMotor` will have the `real_motor` as eg: `th1` and the `pseudo_motor` as `th1_c` **OR** `th1_si` - depending on the material? 
(I have a version of this OffsetMotor that is a PsedoPositioner with an IMS - motor and PseudosingleInterface as the offsets - either `OFF_Si` or `OFF_C`)

Maybe I'm still getting the OffsetMotor architecture wrong, but based on what I understand, we'll have to define these 4 motors at init time, including their `OFF_C` or `OFF_Si` PVs correct?

Something like this:
```
    # thata movement
    th2 = FCpt(OffsetMotor, '{self._th2_offset_prefix}',
               motor_prefix='{self._m_prefix}:MON:MMS:13',
               add_prefix=('prefix', 'motor_prefix'),
               kind='normal', doc='LOM Xtal2 Theta')
```

where `self._th2_offset_prefix` is either `f'{self._prefix}:TH2:OFF_C'` **OR** this one `f'{self._prefix}:TH2:OFF_Si'` 


If I have to do this there is no way for me to know at `__init__` time if we're dealing with `Si` or `C` ... are there ways around this?

Reply:
That's definitely one way to do it. With the offset motor approach (which I had completely forgotten about in my previous comment) you'd need to include both sets of offset motors and decide which ones to link to the energy dynamically based on the Si/C states.

The other way is to drop the offset motors and handle the offset PVs directly in the energy calculations.